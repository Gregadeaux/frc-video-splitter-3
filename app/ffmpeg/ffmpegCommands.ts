import FfmpegCommand, { setFfmpegPath } from 'fluent-ffmpeg';
import { IpcMainEvent } from 'electron';
import moment from 'moment';
import { SplitBlock, SplitDetails, SplitFixedDetails } from "../features/splitter/splitterSlice";

const ffmpegPath = require('@ffmpeg-installer/ffmpeg').path.replace(
  'app.asar',
  'app.asar.unpacked'
);

setFfmpegPath(ffmpegPath);

const progressReportRateMs = 1000;

export function checkVersion(): void {
  console.log(`ffmpegPath: ${ffmpegPath}`);
}

export const buildComplexFilter = (blocks: SplitBlock[]): string[] => {
  const out: string[] = [];
  let concat = '';
  for (let idx = 0; idx < blocks.length; idx += 1) {
    const blk = blocks[idx];
    out.push(`[0:v]trim=${blk.startSeconds}:${blk.startSeconds+blk.durationSeconds},setpts=PTS-STARTPTS[v${idx}]`);
	  out.push(`[0:a]atrim=${blk.startSeconds}:${blk.startSeconds+blk.durationSeconds},asetpts=PTS-STARTPTS[a${idx}]`);
    concat += `[v${idx}][a${idx}]`;
  }
  out.push(`${concat}concat=n=${blocks.length}:v=1:a=1[out]`);
  return out;
};

interface FfmpegFluentProgressData {
  frames: number; // 5245,
  currentFps: number; // 41,
  currentKbps: number; // 3635.6,
  targetSize: number; // 77683,
  timemark: string; // '00:02:55.04',
  percent: number; // 0.730807798968402
}

export async function splitFixedLength(
  event: IpcMainEvent,
  details: SplitFixedDetails
): Promise<null> {
  let lastProgressSent = moment.now();

  return new Promise((resolve, reject) => {
    FfmpegCommand(details.inputFile)
      .on('start', (ffmpegCommand) => {
        console.log(`Command is : ${ffmpegCommand}`);
        event.reply('split-start', {
          matchKey: details.matchKey,
          ffmpegCommand,
        });
      })
      .on('progress', (data: FfmpegFluentProgressData) => {
        const msSinceLastUpdate = moment.now() - lastProgressSent;
        if (msSinceLastUpdate < progressReportRateMs) return;
        console.log(data);

        event.reply('split-progress', {
          matchKey: details.matchKey,
          percent: data.percent * 100,
        });

        lastProgressSent = moment.now();
      })
      .on('end', () => {
        event.reply('split-end', { matchKey: details.matchKey });
        resolve();
      })
      .on('error', (error) => {
        reject(error);
      })
      .complexFilter(buildComplexFilter(details.blocks))
      .outputOptions([
        '-map [out]',
        // '-threads 3',
        // '-vcodec copy',
        // '-acodec copy',
      ])
      .output(`${details.outputFile}`)
      .run();
  });
}
