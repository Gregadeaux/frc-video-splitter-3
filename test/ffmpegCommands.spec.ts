import { buildComplexFilter } from '../app/ffmpeg/ffmpegCommands';

describe('Video Split Component', () => {
  it('test buildComplexFilter', () => {
    const blocks = [
      { startSeconds: 60, durationSeconds: 5 },
      { startSeconds: 120, durationSeconds: 5 },
    ];
    const filter = buildComplexFilter(blocks);

    expect(filter).toMatchSnapshot();
    console.log('filter', filter);
  });

  it('reduces duration', () => {
    const blocks = [
      { startSeconds: 60, durationSeconds: 5 },
      { startSeconds: 120, durationSeconds: 5 },
    ];

    console.log(blocks.map(b=>b.durationSeconds).reduce((prev, cur)=>prev+cur, 0))
  })
});
