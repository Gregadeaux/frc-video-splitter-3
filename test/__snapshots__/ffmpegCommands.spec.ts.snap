// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video Split Component test buildComplexFilter 1`] = `
Array [
  "[0:v]trim=60:65,setpts=PTS-STARTPTS[v0]",
  "[0:a]atrim=60:65,asetpts=PTS-STARTPTS[a0]",
  "[0:v]trim=120:125,setpts=PTS-STARTPTS[v1]",
  "[0:a]atrim=120:125,asetpts=PTS-STARTPTS[a1]",
  "[v0][a0][v1][a1]concat=n=2:v=1:a=1[out]",
]
`;
